dwave/plugins/sklearn/nearest_neighbors.pyx

```diff
@@ -0,0 +1,117 @@
+    cdef np.ndarray[unsigned long, ndim = 2] assignment_counts_x_univariate = np.zeros((sub_index_size, n_variables), dtype=np.uint)
+    cdef np.ndarray[unsigned long, ndim = 1] assignment_counts_y = np.zeros((sub_index_size), dtype=np.uint)
+    cdef np.ndarray[unsigned long, ndim = 5] assignment_counts_joint = np.zeros((sub_index_size, sub_index_size, sub_index_size, n_variables, n_variables), dtype=np.uint)
+    cdef np.ndarray[unsigned long, ndim = 3] assignment_counts_joint_univariate = np.zeros((sub_index_size, sub_index_size, n_variables), dtype=np.uint)
+
+    cdef np.ndarray[double, ndim = 2] mi_matrix = np.zeros((n_variables, n_variables))
+
+    for variable_1 in range(n_variables):
+        for variable_2 in range(n_variables):
+            if variable_1 > variable_2:
+                for obs in range(n_obs):
+                    assignment_counts_x[assignments[obs, variable_1],assignments[obs, variable_2],variable_1, variable_2] += 1
+                    assignment_counts_joint[assignments[obs, variable_1],assignments[obs, variable_2], assignments[obs, -1],variable_1, variable_2] += 1
+
+        for obs in range(n_obs):
+            assignment_counts_x_univariate[assignments[obs, variable_1], variable_1] += 1
+            assignment_counts_joint_univariate[assignments[obs, variable_1], assignments[obs, -1], variable_1] += 1
+
+    for obs in range(n_obs):
+        assignment_counts_y[assignments[obs, -1]] += 1
+    
+    for variable_1 in range(n_variables):
+        for variable_2 in range(n_variables):
+            if variable_1 > variable_2:
+                for i in range(sub_index_size): 
+                    for j in range(sub_index_size): 
+                        for k in range(sub_index_size):
+                            if assignment_counts_joint[i,j,k,variable_1, variable_2] > 0:
+                                ijk_term = (<float> assignment_counts_joint[i,j,k,variable_1, variable_2])
+                                ijk_term *= log(ijk_term) - log(assignment_counts_x[i,j,variable_1, variable_2]) - log(assignment_counts_y[k]) + log_n_obs
+                                mi_matrix[variable_1, variable_2] += ijk_term
+                                mi_matrix[variable_2, variable_1] += ijk_term
+        for i in range(sub_index_size): 
+            for k in range(sub_index_size):
+                if assignment_counts_joint_univariate[i,k,variable_1] > 0:
+                    ik_term = (<float> assignment_counts_joint_univariate[i,k,variable_1])
+                    ik_term *= log(ik_term) - log(assignment_counts_x_univariate[i,variable_1]) - log(assignment_counts_y[k]) + log_n_obs
+                    mi_matrix[variable_1, variable_1] += ik_term
+    
+    return mi_matrix/n_obs
+
+cdef assign_nn(np.ndarray[np.float_t,ndim = 2] X,
+                long sub_index_size, 
+                short seed = 12121):
```

Comment:
In general, it's preferred to to make the default seed `None`.

Also, in this case you're better off _not_ specifying a C type for `seed`. This is because it's only passed to a python function, `np.random.default_rng(seed)`, below. So if it starts out as a C-type, it will have to be converted to a Python type first. Of course, in this case the performance hit is so minor as to be pointless... but the other advantage of using `object` here rather than `short` is it supports `None` or another rng.